### PR TITLE
QPPCT-406: verify cpc scenario correctness

### PR DESCRIPTION
**Pull request by clydet**
### Information
- JIRA story QPPCT-406

### Changes proposed in this PR:
- puts a framework in place that disambiguates negative scenarios for CPC+ acceptance tests

### Checklist 
- [x] All JUnit tests pass (`mvn clean verify`).
- [x] New unit tests written to cover new functionality.
- [ ] Added and updated JavaDocs for non-test classes and methods.
- [x] No local design debt. Do you feel that something is "ugly" after your changes?
- [ ] Updated documentation (`README.md`, etc.) depending if the changes require it.

### CODE REVIEW

File: converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java [hunk 1]
```diff
@@ -60,18 +72,64 @@ void testCpcPlusFileFailures() throws IOException {
...
+		assertWithMessage("The fixture file is not representative of the failures directory")
+				.that(getXml(FAILURE).count())
+				.isEqualTo(fixtureValues.size());
 		assertThat(successesThatShouldBeErrors).isEmpty();
 	}
 
+	private void verifyOutcome(String filename, List<Detail> details) {
+		CPCAcceptanceFixture expectedErrors = fixtureValues.get(filename);
+
+		System.out.println("Verifying scenario " + filename +
```
Comment (sfradkin) [inline converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java]: Logger please

File: converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java [hunk 2]
```diff
@@ -21,23 +24,32 @@
...
+				new TypeReference<Map<String, CPCAcceptanceFixture>>() { };
+		fixtureValues = JsonHelper.readJson(FAILURE_FIXTURE, ref);
 	}
 
 	@AfterAll
 	static void resetApmIds() {
 		ApmEntityIds.setApmDataFile(ApmEntityIds.DEFAULT_APM_ENTITY_FILE_NAME);
 	}
 
+	@Disabled
```
Comment (sfradkin) [inline converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java]: why are we disabling the success verification?

File: converter/src/main/java/gov/cms/qpp/conversion/model/error/ErrorCode.java [hunk 3]
```diff
@@ -15,7 +15,7 @@
 	UNEXPECTED_ERROR(3, "Unexpected exception occurred during conversion"),
 	UNEXPECTED_ENCODE_ERROR(4, "Unexpected exception occured during encoding"),
 	NOT_VALID_QRDA_DOCUMENT(5, "The file is not a QRDA-III XML document"),
-	MEASURE_GUID_MISSING(6, "The measure reference results must have a measure GUID"),
+	MEASURE_GUID_MISSING(6, "The measure reference results must have a recognized measure GUID"),
```
Comment (sfradkin) [inline converter/src/main/java/gov/cms/qpp/conversion/model/error/ErrorCode.java]: minor issue, but please update ERROR_MESSAGES.md with this message change

File: converter/src/test/java/gov/cms/qpp/acceptance/cpc/CPCAcceptanceFixture.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package gov.cms.qpp.acceptance.cpc;
+
+import java.util.List;
+
+public class CPCAcceptanceFixture {
```
Comment (sfradkin) [inline converter/src/test/java/gov/cms/qpp/acceptance/cpc/CPCAcceptanceFixture.java]: We don't normally javadoc test classes, but could you put a bit of documentation here to explain a bit about what this class does?

File: converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java [hunk 5]
```diff
@@ -60,18 +72,64 @@ void testCpcPlusFileFailures() throws IOException {
...
 		assertThat(successesThatShouldBeErrors).isEmpty();
 	}
 
+	private void verifyOutcome(String filename, List<Detail> details) {
+		CPCAcceptanceFixture expectedErrors = fixtureValues.get(filename);
+
+		System.out.println("Verifying scenario " + filename +
+				", expected errors match the following actual errors. " + details);
+
+		if (expectedErrors.isStrict()) {
```
Comment (sfradkin) [inline converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java]: Don't we always want to be strict about this?  I saw in the meta-data that we are not strict in some scenarios.

File: converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java [hunk 6]
```diff
@@ -60,18 +72,64 @@ void testCpcPlusFileFailures() throws IOException {
...
+				//TODO remove this once all error scenarios are validated
+				if (fixtureValues.get(entry.toFile().getName()) == null) {
+					return false;
+				}
 				try {
 					converter.transform();
 					return true;
 				} catch (TransformException expected) {
+					//runnning conversions on individual files
+					List<Detail> details = expected.getDetails().getErrors().get(0).getDetails();
```
Comment (sfradkin) [inline converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java]: Is there any way to refactor how the `TransformException` works with retrieving the details?  This is a long stream of function calls that's not real easy to understand.

File: converter/src/test/java/gov/cms/qpp/acceptance/cpc/FixtureErrorData.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
+package gov.cms.qpp.acceptance.cpc;
+
+import gov.cms.qpp.conversion.model.error.ErrorCode;
+
+import java.util.List;
+
+public class FixtureErrorData {
```
Comment (sfradkin) [inline converter/src/test/java/gov/cms/qpp/acceptance/cpc/FixtureErrorData.java]: Another case where just a bit of documentation will be helpful in understanding the data this class is used to represent.

File: converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java (hunk 5 above)
Comment (halprin) [reply converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java]: We don't want to always be strict.  In the future, other validations may come along that makes one of these CPC+ XML files generate a new validation error and that would then throw off the strict assert and fail the test.  _But_, we shouldn't need to care that this XML file now has a new additional validation error on it.  We should only care that it still fails in the original way.

File: converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java (hunk 1 above)
Comment (halprin) [reply converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java]: Logger in a test?

File: converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java (hunk 5 above)
Comment (sfradkin) [reply converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java]: But is it a bad thing to always be strict?  It ensures that we've enumerated all the errors that could occur with the file.  If a new validation causes a file to throw an error, then add it to the list of expected validation errors.  That way we know everything that's expected and that there's no issues with unexpected interactions.

Comment (halprin) [reply converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java]: I think it is bad thing, actually.  These CPC+ files were added to test the existence of specific errors in them, so we should only test for those specific errors and not any possible additional ones.

In past projects I have been on, we decided to be strict.  This led to many tests failing when small changes were introduced.  At best, this creates unnecessary churn and a groaning engineer, and at worst, creates an environment where the engineer is overwhelmed with all the test failures and just decides to fix the test by grabbing the actual results and making them the expected results (and ta-da, the test passes now).

File: converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java (hunk 2 above)
Comment (clydet) [reply converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java]: oops, meant to reinstate that

File: converter/src/test/java/gov/cms/qpp/acceptance/cpc/CPCAcceptanceFixture.java (hunk 4 above)
Comment (clydet) [reply converter/src/test/java/gov/cms/qpp/acceptance/cpc/CPCAcceptanceFixture.java]: kk

File: converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java (hunk 6 above)
Comment (clydet) [reply converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java]: Not sure we have a usage for this outside of this test class. Will give it a look though.

Comment (sfradkin) [reply converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java]: sounds good.  if it's not being used anywhere else I'm ok with it.

File: converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java (hunk 1 above)
Comment (sfradkin) [reply converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java]: ah, right... it's a test.  still seems weird to see a raw `System.out`

Comment (clydet) [reply converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java]: ok

File: converter/src/main/java/gov/cms/qpp/conversion/validate/CpcQualityMeasureIdValidator.java [hunk 8]
```diff
@@ -17,6 +17,7 @@
 import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
```
Comment (FlexionSonarQube) [inline converter/src/main/java/gov/cms/qpp/conversion/validate/CpcQualityMeasureIdValidator.java]: MINOR Remove this unused import 'java.util.Optional'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by FlexionSonarQube** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: converter/src/test/resources/cpc_plus/failure/CPCPlus_IncNumTotalMeasures_SampleQRDA-III.xml [hunk 9]
```diff
@@ -92,7 +92,7 @@
 	  <participant typeCode="LOC">
       <associatedEntity classCode="SDLOC">
          <id root="2.16.840.1.113883.3.249.5.1"
-            extension="T2OR1234"
+            extension="DogCow"
```
Comment (halprin) [inline converter/src/test/resources/cpc_plus/failure/CPCPlus_IncNumTotalMeasures_SampleQRDA-III.xml]: Moof!  Also, `TestApmEntityId` should work too, given that it is in `test_apm_entity_ids.json`.

Comment (clydet) [reply converter/src/test/resources/cpc_plus/failure/CPCPlus_IncNumTotalMeasures_SampleQRDA-III.xml]: true, but I prefer the inter-species hybrid

File: converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java (hunk 1 above)
Comment (omnivion) [reply converter/src/test/java/gov/cms/qpp/acceptance/cpc/CpcPlusAcceptanceTest.java]: Agreed. JUnit 5 has a way to handle this through parameterized tests. You can generate test data in a Stream.

**Pull request merged**